include/pressio/ops/tpetra/ops_min_max.hpp

```diff
@@ -0,0 +1,123 @@
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Francesco Rizzi (fnrizzi@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef OPS_TPETRA_OPS_MIN_MAX_HPP_
+#define OPS_TPETRA_OPS_MIN_MAX_HPP_
+
+namespace pressio{ namespace ops{
+
+template <typename T>
+::pressio::mpl::enable_if_t<
+  // TPL/container specific
+    (::pressio::is_vector_tpetra<T>::value
+  || ::pressio::is_multi_vector_tpetra<T>::value)
+  // scalar compatibility
+  && (std::is_floating_point<typename ::pressio::Traits<T>::scalar_type>::value
+   || std::is_integral<typename ::pressio::Traits<T>::scalar_type>::value),
+  typename ::pressio::Traits<T>::scalar_type
+  >
+max(const T & obj)
+{
+  assert(::pressio::ops::extent(obj, 0) > 0);
+  assert(!::pressio::is_multi_vector_tpetra<T>::value
+       || ::pressio::ops::extent(obj, 1) > 0);
+
+  // get local minimum
+  using sc_t = typename ::pressio::Traits<T>::scalar_type;
+  auto local = obj.getLocalViewDevice(Tpetra::Access::ReadOnly);
+  sc_t max_local = ::pressio::ops::max(local);
+
+  // reduce over ranks to get global minimum
+  sc_t max_global = max_local;
+  Teuchos::reduceAll<int, sc_t>(
+    *obj.getMap()->getComm(),
+    Teuchos::REDUCE_MAX,
+    1,
+    &max_local,
+    &max_global
+    );
+
+  return max_global;
+}
+
+template <typename T>
+::pressio::mpl::enable_if_t<
+  // TPL/container specific
+    (::pressio::is_vector_tpetra<T>::value
+  || ::pressio::is_multi_vector_tpetra<T>::value)
+  // scalar compatibility
+  && (std::is_floating_point<typename ::pressio::Traits<T>::scalar_type>::value
+   || std::is_integral<typename ::pressio::Traits<T>::scalar_type>::value),
+  typename ::pressio::Traits<T>::scalar_type
+  >
+min(const T & obj)
+{
+  assert(::pressio::ops::extent(obj, 0) > 0);
+  assert(!::pressio::is_multi_vector_tpetra<T>::value
+       || ::pressio::ops::extent(obj, 1) > 0);
+
+  // get local minimum
+  using sc_t = typename ::pressio::Traits<T>::scalar_type;
+  auto local = obj.getLocalViewDevice(Tpetra::Access::ReadOnly);
+  sc_t min_local = ::pressio::ops::min(local);
+
+  // reduce over ranks to get global minimum
+  sc_t min_global = min_local;
+  Teuchos::reduceAll<int, sc_t>(
+    *obj.getMap()->getComm(),
+    Teuchos::REDUCE_MIN,
+    1,
+    &min_local,
+    &min_global
+    );
+
+  return min_global;
```

Comment:
@fnrizzi
Could you please confirm that this method is correct ?
I've found similar global reduction (based on `Teuchos::reduceAll()`) in [Tpetra Tutorial](https://docs.trilinos.org/dev/packages/tpetra/doc/html/Tpetra_Lesson02.html)

Reply:
`Epetra_MultiVector` implements [`MinValue()`](https://docs.trilinos.org/dev/packages/epetra/doc/html/classEpetra__MultiVector.html#a29c2e3555f1004c01e9551ff71c2fe54) and [`MaxValue()`](https://docs.trilinos.org/dev/packages/epetra/doc/html/classEpetra__MultiVector.html#a9a09de139ff1e36842649a865bf7d3ff), which I can't find in [`Tpetra::MultiVector`](https://docs.trilinos.org/dev/packages/tpetra/doc/html/classTpetra_1_1MultiVector.html),
so I just wandered if direct reduction is our only option here ?

Reply:
i don't know if it is only option, but i guess the kernel is correct, did you chec in the tests?

Reply:
Yes, it's covered and [the tests](https://github.com/Pressio/pressio/pull/554/files#diff-04402219f3a8ced2accf5d97363e5ed97b4e78c126edc917add8d47ade9a94bb) take the global reduction into account.